byzcoin/transaction.go

```diff
@@ -423,8 +425,35 @@ func decodeStateChangeBody(buf []byte) (StateChangeBody, error) {
 // StateChanges hold a slice of StateChange
 type StateChanges []StateChange
 
+// Len returns the number of state changes
+func (scs StateChanges) Len() int {
+	return len(scs)
+}
+
+// Less returns true if i is before j (by instance and then by version)
+func (scs StateChanges) Less(i, j int) bool {
+	if bytes.Equal(scs[i].InstanceID, scs[j].InstanceID) {
+		return scs[i].Version < scs[j].Version
+	}
+
+	return bytes.Compare(scs[i].InstanceID, scs[j].InstanceID) < 0
+}
+
+// Swap swaps two elements
+func (scs StateChanges) Swap(i, j int) {
+	scs[i], scs[j] = scs[j], scs[i]
+}
+
 // Hash returns the sha256 of all stateChanges
 func (scs StateChanges) Hash() []byte {
+	if !sort.IsSorted(scs) {
+		// An order needs to be defined so that we can compute
+		// a hash from an arbitrary list of state changes
+		// (i.e. the cache and the history storage have different
+		// ordering which is producing a different hash)
+		sort.Sort(scs)
+	}
```

Comment:
depending on how you use this function, i have a suspicion that sorting StateChange will have a conflict with the SignerCounter change that just got merged. because if one user submits a lot of instructions, which results in a lot of state changes, then they must happen in the same order, otherwise the SignerCounter won't be monotonically increasing. I guess we'll tell only when we try to merge.

Reply:
Ok I will have a particular look at that when rebasing